### PR TITLE
llama : apply classifier-free guidance to logits directly

### PR DESCRIPTION
**Pull request by dfriehs**
With the new `llama_batch` API calls to `llama_sample_classifier_free_guidance` might return garbage as it currently only samples logits at position 0.

However, instead of adding a new parameter denoting the position that logits should be sampled from, I propose to change the sampling method to operate on logits directly; moving the call above the `llama_token_data_array` creation:
- CFG is always used in the beginning, asserting that the candidates have neither been sorted nor that some have been excluded already. With the new method signature this is more explicit.
- The code path can be simplified a lot as classifier free guidance operates on logits already and would have to create a temporary vector of logits otherwise.
- Call site is now able to use logits for any position.

`llama_sampling_sample` has been changed to use logits from the same position as for the main context as it is assumed that the guidance context has received the exact same batches as the main context.

I welcome any feedback or request for changes.

### CODE REVIEW

**Review by ggerganov — APPROVED** [review]
Good change - this is a better API
